### PR TITLE
Add async TCP echo example

### PR DESCRIPTION
**Pull request by FranzBusch**
# Motivation
We introduced new async APIs to our bootstraps that make it easy to use NIO from Swift Concurrency. We should provide examples showing how to use those new APIs. In the future, we want to make those examples the primary ones and remove the non-async based ones.

# Modification
This PR introduces two new targets: `NIOTCPEchoServer` and `NIOTCPEchoClient`. Both targets are executable and provide either the server or client piece of the example.

# Result
New and shiny async example

### CODE REVIEW

File: Sources/NIOTCPEchoClient/Client.swift
Comment (glbrntt) [inline Sources/NIOTCPEchoClient/Client.swift:47]: why do we sleep?

File: Sources/NIOTCPEchoClient/Client.swift
Comment (glbrntt) [inline Sources/NIOTCPEchoClient/Client.swift:39]: Pretty sure this is a client ;)

File: Sources/NIOTCPEchoClient/Client.swift
Comment (glbrntt) [inline Sources/NIOTCPEchoClient/Client.swift:85]:
```suggestion
}
#else
@main
```

File: Sources/NIOTCPEchoServer/Server.swift
Comment (glbrntt) [inline Sources/NIOTCPEchoServer/Server.swift:62]: I think we should elaborate on this a bit because it makes it sound like task groups leak! We don't leak memory if we use a task group, we just don't free it until the task group exits or we consume the results.

File: Sources/NIOTCPEchoClient/Client.swift
Comment (weissi) [inline Sources/NIOTCPEchoClient/Client.swift:37]: Sure, in this case you don't have to shut it down but why leak stuff if it's trivial to not leak?

`defer { try! elg.syncShutdownGracefully() }`

File: Sources/NIOTCPEchoClient/Client.swift
Comment (weissi) [inline Sources/NIOTCPEchoClient/Client.swift:78]: The string conversion here is incorrect. You can receive in arbitrarily sized chunks. So either you frame or you write each chunk as bytes (and crucially not String).

File: Sources/NIOTCPEchoServer/Server.swift
Comment (weissi) [inline Sources/NIOTCPEchoServer/Server.swift:37]: unnecessary (but in this case benign leak). But be mindful that this code will end up copy&pasted potentially into functions that get called multiple times.

Comment (weissi) [reply Sources/NIOTCPEchoServer/Server.swift:37]: If you really want a singleton, it should be a lazy global variable and named appropriately.

File: Sources/NIOTCPEchoServer/Server.swift
Comment (weissi) [inline Sources/NIOTCPEchoServer/Server.swift:81]: same issue here, the printing is incorrect as we're not framing

File: Sources/NIOTCPEchoClient/Client.swift
Comment (weissi) [reply Sources/NIOTCPEchoClient/Client.swift:37]: So moving this into a global variable and naming it `globalSingletonGroup` or so would be much better because copy & paste wouldn't lead to super bad leaks. The code as is is just incorrect because `eventLoopGroup` will get `deinit`ed and it's illegal to `deinit` this without shutting it down (don't think that's enforced today though).

Thinking about it more: I've long come to the conclusion that NIO needs to provides singletons for users who don't need tight control and I think https://github.com/apple/swift-nio/pull/2471 is the answer :)

**Review by weissi — CHANGES_REQUESTED** [review]
Just leaking an `EventLoopGroup` in a _function_ is illegal (will get deinited without shutdown which is a lifecycle violation)

File: Sources/NIOTCPEchoClient/Client.swift
Comment (FranzBusch) [reply Sources/NIOTCPEchoClient/Client.swift:47]: If a user runs this in Xcode he gets a couple of seconds to see the output. Otherwise it is just gone immediately.

Comment (glbrntt) [reply Sources/NIOTCPEchoClient/Client.swift:47]: Oof, that's pretty annoying. Could you add a comment explaining that and that it's not actually necessary? Examples get copy-pasted a lot so we don't want users to actually do this!

File: Sources/NIOTCPEchoClient/Client.swift
Comment (FranzBusch) [reply Sources/NIOTCPEchoClient/Client.swift:37]: Holding this PR until we land the singleton work. Thanks @weissi!

File: Sources/NIOTCPEchoClient/Client.swift
Comment (FranzBusch) [reply Sources/NIOTCPEchoClient/Client.swift]: I added a newline based decoder/encoder to both pipelines. I think that is actually good to have in an example since if somebody copies it over they are aware of the need for framing right away.

**Comment by FranzBusch** [comment]
@weissi I updated this PR to use your newly landed singletons.

File: Sources/NIOTCPEchoClient/Client.swift
Comment (weissi) [reply Sources/NIOTCPEchoClient/Client.swift:39]: Yeah, can we not do that? Let's find a better way that demonstrates this without just delaying it here...

File: Sources/NIOTCPEchoClient/Client.swift
Comment (weissi) [inline Sources/NIOTCPEchoClient/Client.swift:96]:
```suggestion
    private let newLine = UInt8(ascii: "\n")
```

File: Sources/NIOTCPEchoClient/Client.swift
Comment (FranzBusch) [reply Sources/NIOTCPEchoClient/Client.swift:39]: I mean I am happy to just not do this.

File: Sources/NIOTCPEchoClient/Client.swift
Comment (weissi) [inline Sources/NIOTCPEchoClient/Client.swift:93]: could we please not use `*Unsafe*` in the examples. `buffer.readableBytesView.firstIndex(of: UInt8(ascii: "\n"))` gives you what you need.

Comment (weissi) [reply Sources/NIOTCPEchoClient/Client.swift:93]: or probably something like

```
let readableBytes = buffer.readableBytesView

if let firstLine = readableBytes.firstIndex(of: UInt8(ascii: "\n")).map({ readableBytes[readableBytes.startIndex ..< $0] }) {
    buffer.moveReaderIndex(forwardBy: firstLine.count)
    context.fireChannelRead(self.wrapInboundOut(ByteBuffer(firstLine.dropLast() /* kill newline */)))
    return .continue
} else {
    return .needMoreData
}
```

File: Sources/NIOTCPEchoServer/Server.swift
Comment (weissi) [inline Sources/NIOTCPEchoServer/Server.swift:104]: same here, unnecessary Unsafe

File: Sources/NIOTCPEchoServer/Server.swift
Comment (weissi) [inline Sources/NIOTCPEchoServer/Server.swift:52]: `.init` is pretty ugly, could we get a type name?

File: Sources/NIOTCPEchoServer/Server.swift
Comment (weissi) [inline Sources/NIOTCPEchoServer/Server.swift:47]: all these `try`s should be `try!`s, they can't fail unless programmer error, right?

Comment (FranzBusch) [reply Sources/NIOTCPEchoServer/Server.swift:47]: They could throw if the channel is closed while we add the handler. I don't wanna show adopters a `try!` when it could fail in rare cases and I don't think the `try` is too bad.

**Review by glbrntt — APPROVED** [review]

**Review by weissi — APPROVED** [review]
lgtm

**Pull request merged**
